core/src/main/java/com/google/instrumentation/trace/ProbabilitySampler.java

```diff
@@ -0,0 +1,92 @@
+  // probability range.  Using the low bits of the traceId also ensures that systems that only use
+  // 64 bit ID's will also work with this sampler.
+  private final long idUpperBound;
+  private final double probability;
+
+  /**
+   * Create a new {@link ProbabilitySampler}. The probability of sampling a trace is equal to that
+   * of the specified probability.
+   *
+   * @param probability The desired probability of sampling. Must be within [0.0, 1.0].
+   * @throws IllegalArgumentException if {@code probability} is out of range
+   */
+  public ProbabilitySampler(double probability) {
+    this.probability = probability;
+    if (probability < 0.0 || probability > 1.0) {
+      throw new IllegalArgumentException("probability must be in range [0.0, 1.0]");
+    }
+    // Special case the limits, to avoid any possible issues with lack of precision across
+    // double/long boundaries. For probability == 0.0, we use Long.MIN_VALUE as this guarantees
+    // that we will never sample a trace, even in the case where the id == Long.MIN_VALUE, since
+    // Math.Abs(Long.MIN_VALUE) == Long.MIN_VALUE.
+    if (probability == 0.0) {
+      idUpperBound = Long.MIN_VALUE;
+    } else if (probability == 1.0) {
+      idUpperBound = Long.MAX_VALUE;
+    } else {
+      idUpperBound = (long) (probability * Long.MAX_VALUE);
+    }
+  }
+
+  /**
+   * Return the probability this this {@link Sampler} will return true for {@link
+   * Sampler#shouldSample}
+   *
+   * @return the probability that this {@code Sampler} will return true for {@code
+   *     Sampler.shouldSample()}
+   */
+  public double getProbability() {
+    return probability;
+  }
+
+  @Override
+  public boolean shouldSample(
+      @Nullable SpanContext parentContext,
+      boolean remoteParent,
+      TraceId traceId,
+      SpanId spanId,
+      String name,
+      @Nullable List<Span> parentLinks) {
+    checkNotNull(traceId, "traceId");
+    checkNotNull(name, "name");
+    // TODO(bdrutu): Fix this to use TraceId#compareTo.
+    BigInteger traceIdNum = new BigInteger(traceId.getBytes());
+    // Always sample if we are within probability range. This is true even for child spans (that
+    // may have had a different sampling decision made) to allow for different sampling policies,
+    // and dynamic increases to sampling probabilities for debugging purposes.
+    // Note use of '<' for comparison. This ensures that we never sample for probability == 0.0,
+    // while allowing for a (very) small chance of *not* sampling if the id == Long.MAX_VALUE.
+    // This is considered a reasonable tradeoff for the simplicity/performance requirements (this
+    // code is executed in-line for every Span creation).
+    if (Math.abs(traceIdNum.longValue()) < idUpperBound) {
+      return true;
+    }
+    // Also enable sampling if parent was sampled.
+    return (parentContext != null && parentContext.getTraceOptions().isSampled());
```

Comment:
We can check this first because this will always return true if the parent is sampled and we save cycles.